src/ui.rs

```diff
@@ -0,0 +1,335 @@
+use macroquad::prelude::*;
+
+use std::{thread::sleep, time::Duration};
+
+const SQUARES: i16 = 32;
+const TEXT_SIZE: f32 = 25.;
+const TEXT_POS_X: f32 = 0.2;
+const TEXT_POS_Y: f32 = 0.9;
+const TITLE_SIZE: f32 = 40.;
+const TITLE_POS_X: f32 = 0.3;
+const TITLE_POS_Y: f32 = 0.05;
+const GENERATIONS_POS_X: f32 = 0.1;
+const GENERATIONS_POS_Y: f32 = 0.1;
+const DEFAULT_SEED: [(i32, i32); 8] = [
+    (14, 13),
+    (14, 14),
+    (14, 15),
+    (15, 13),
+    (15, 15),
+    (16, 13),
+    (16, 14),
+    (16, 15),
+];
+
+/// This function displays the user interface for the Game of Life, including rendering the game screen,
+/// handling user input, and switching between different modes (playing, paused, edit).
+///
+/// # Parameters
+///
+/// - `game`: The current game state.
+/// - `game_running`: A flag indicating whether the game is running.
+/// - `game_paused`: A flag indicating whether the game is paused.
+/// - `edit_mode`: A flag indicating whether the game is in edit mode.
+/// - `generations_passed`: The number of generations that have passed in the game.
+///
+async fn display_ui(
+    game: &mut Game,
+    game_running: &mut bool,
+    game_paused: &mut bool,
+    edit_mode: &mut bool,
+    generations_passed: &mut i32,
+) {
+    let (offset_x, offset_y, sq_size) = render_screen(game, generations_passed);
+
+    if *game_paused {
+        if *edit_mode {
+            run_edit_mode(game, &offset_x, &offset_y, &sq_size);
+        } else {
+            run_paused_mode();
+        }
+    } else {
+        run_playing_mode(game, generations_passed);
+    }
+
+    handle_commands(game, game_running, game_paused, edit_mode);
+
+    next_frame().await;
+}
+
+/// Renders the game screen with grid lines and the live cells.
+///
+/// # Parameters
+///
+/// - `game`: The current game state.
+/// - `generations_passed`: The number of generations that have passed in the game.
+///
+fn render_screen(game: &Game, generations_passed: &i32) -> (f32, f32, f32) {
+    clear_background(LIGHTGRAY);
+    let screen_width = screen_width();
+    let screen_height = screen_height();
+
+    let game_size = screen_width.min(screen_height);
+    let offset_x = (screen_width - game_size) / 2. + 10.;
+    let offset_y = (screen_height - game_size) / 2. + 10.;
+    let sq_size = (screen_height - offset_y * 2.) / SQUARES as f32;
+
+    draw_rectangle(offset_x, offset_y, game_size - 20., game_size - 20., WHITE);
+
+    for i in 1..SQUARES {
+        draw_line(
+            offset_x,
+            offset_y + sq_size * i as f32,
+            screen_width - offset_x,
+            offset_y + sq_size * i as f32,
+            2.,
+            LIGHTGRAY,
+        );
+    }
+
+    for i in 1..SQUARES {
+        draw_line(
+            offset_x + sq_size * i as f32,
+            offset_y,
+            offset_x + sq_size * i as f32,
+            screen_height - offset_y,
+            2.,
+            LIGHTGRAY,
+        );
+    }
+
+    for (x, y) in game.live_cells() {
```

Comment:
For not breaking encapsulation this is great, but for a performance perspective I think doing a `collect` every iteration could make it a little laggy, to have the best of both worlds, I'd recommend having a function that receives a closure, something like `game.live_cells_do<F: Fn((i32, i32))>(&self, closure: F)`

Reply:
Done!